test/extended/images/mongodb_replica.go

```diff
@@ -0,0 +1,126 @@
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	g "github.com/onsi/ginkgo"
+	o "github.com/onsi/gomega"
+
+	exutil "github.com/openshift/origin/test/extended/util"
+	"github.com/openshift/origin/test/extended/util/db"
+)
+
+var _ = g.Describe("[images][mongodb] openshift mongodb replication", func() {
+	defer g.GinkgoRecover()
+
+	const (
+		templatePath         = "https://raw.githubusercontent.com/openshift/mongodb/master/2.4/examples/replica/mongodb-clustered.json"
+		deploymentConfigName = "mongodb"
+		insertCmd            = `db.bar.save({ "status": "passed" })`
+	)
+
+	const (
+		expectedReplicasAfterDeployment = 3
+		expectedReplicasAfterScalingUp  = expectedReplicasAfterDeployment + 2
+	)
+
+	oc := exutil.NewCLI("mongodb-replica", exutil.KubeConfigPath()).Verbose()
+
+	g.Describe("creating from a template", func() {
+		g.It(fmt.Sprintf("should process and create the %q template", templatePath), func() {
+
+			g.By("creating a new app")
+			o.Expect(oc.Run("new-app").Args("-f", templatePath).Execute()).Should(o.Succeed())
+
+			g.By("waiting for the deployment to complete")
+			waitForDeploymentToComplete(oc, deploymentConfigName)
+
+			podNames := waitForNumberOfPodsWithLabel(oc, expectedReplicasAfterDeployment, "mongodb-replica")
+			mongo := db.NewMongoDB(podNames[0])
+
+			g.By(fmt.Sprintf("expecting that replica set have %v members", expectedReplicasAfterDeployment))
+			assertMembersInReplica(oc, mongo, expectedReplicasAfterDeployment)
+
+			g.By("expecting that we can insert a new record on primary node")
+			_, err := mongo.QueryPrimary(oc, insertCmd)
+			o.Expect(err).ShouldNot(o.HaveOccurred())
+
+			g.By("expecting that we can read a record from all members")
+			for _, podName := range podNames {
+				tryToreadFromPod(oc, podName)
+			}
+
+			scale(oc, deploymentConfigName, expectedReplicasAfterScalingUp)
+
+			podNames = waitForNumberOfPodsWithLabel(oc, expectedReplicasAfterScalingUp, "mongodb-replica")
+			mongo = db.NewMongoDB(podNames[0])
+
+			g.By("expecting that scaling replica set up should have more members")
+			assertMembersInReplica(oc, mongo, expectedReplicasAfterScalingUp)
+		})
+	})
+
+})
+
+func waitForDeploymentToComplete(oc *exutil.CLI, deploymentConfigName string) {
+	exutil.WaitForADeployment(
+		oc.KubeREST().ReplicationControllers(oc.Namespace()),
+		deploymentConfigName,
+		exutil.CheckDeploymentCompletedFn,
+		exutil.CheckDeploymentFailedFn,
+	)
+}
+
+func tryToreadFromPod(oc *exutil.CLI, podName string) {
+	// don't include _id field to output because it changes every time
+	const findCmd = "rs.slaveOk(); printjson(db.bar.find({}, {_id: 0}).toArray())"
+
+	fmt.Fprintf(g.GinkgoWriter, "DEBUG: reading record from pod %v\n", podName)
+
+	mongoPod := db.NewMongoDB(podName)
+	result, err := mongoPod.Query(oc, findCmd)
+	o.Expect(err).ShouldNot(o.HaveOccurred())
+	o.Expect(result).Should(o.ContainSubstring(`{ "status" : "passed" }`))
+}
+
+func scale(oc *exutil.CLI, deploymentConfigName string, replicas int) {
```

Comment:
+1 i'd rather see this inline.  it's also harder to trace failures when they happen inside a shared function since you just get the line number and have to go figure out which test was calling it.